libvast/src/system/local_segment_store.cpp

```diff
@@ -0,0 +1,306 @@
+      VAST_DEBUG("{} delegates {} deferred evaluations", self,
+                 self->state.deferred_requests.size());
+      for (auto&& [expr, ids, rp] :
+           std::exchange(self->state.deferred_requests, {}))
+        rp.delegate(static_cast<store_actor>(self), std::move(expr),
+                    std::move(ids));
+    });
+  return {
+    // store
+    [self](query query, ids ids) -> caf::result<atom::done> {
+      VAST_WARN("got a query for some ids");
+      if (!self->state.segment) {
+        auto rp = caf::typed_response_promise<atom::done>();
+        self->state.deferred_requests.emplace_back(query, ids, rp);
+        return rp;
+      }
+      // Special-case handling for "erase"-queries because their
+      // implementation must be different depending on if we operate
+      // in memory or on disk.
+      if (caf::holds_alternative<query::erase>(query.cmd)) {
+        return self->delegate(static_cast<store_actor>(self), atom::erase_v,
+                              ids);
+      }
+      auto slices = self->state.segment->lookup(ids);
+      if (!slices)
+        return slices.error();
+      return handle_lookup(self, query, ids, *slices);
+    },
+    [self](atom::erase, ids xs) -> caf::result<atom::done> {
+      if (!self->state.segment) {
+        // Treat this as an "erase" query for the purposes of storing it
+        // until the segment is loaded.
+        auto rp = caf::typed_response_promise<atom::done>();
+        self->state.deferred_requests.emplace_back(query::make_erase({}), xs,
+                                                   rp);
+        return rp;
+      }
+      auto new_segment = segment::copy_without(*self->state.segment, xs);
+      if (!new_segment) {
+        VAST_ERROR("could not remove ids from segment {}: {}",
+                   self->state.segment->id(), render(new_segment.error()));
+        return new_segment.error();
+      }
+      VAST_ASSERT(self->state.path.has_filename());
+      auto old_path = self->state.path;
+      auto new_path = self->state.path.replace_extension("next");
+      // TODO: If the new segment is empty, we should probably just erase the
+      // file without replacement here.
+      self
+        ->request(self->state.fs, caf::infinite, atom::write_v, new_path,
+                  new_segment->chunk())
+        .then(
+          [seg = std::move(*new_segment), self, old_path,
+           new_path](atom::ok) mutable {
+            std::error_code ec;
+            // Re-use the old filename so that we don't have to write a new
+            // partition flatbuffer with the changed store header as well.
+            std::filesystem::rename(new_path, old_path, ec);
+            if (ec)
+              VAST_ERROR("failed to erase old data {}", seg.id());
+            self->state.segment = std::move(seg);
+          },
+          [](caf::error& err) {
+            VAST_ERROR("failed to flush archive {}", to_string(err));
+          });
+      return atom::done_v;
```

Comment:
Return response promise here and only return `done` after the file was overwritten.